### PR TITLE
refactor: bump datafusion, arrow and parquet

### LINKED ISSUE #25: Bump Arrow, Parquet and DataFusion
**Description**

Bump versions.
Target:
- Arrow: 13.0.0
- Parquet: 13.0.0
- DataFusion: 8.0.0

Current:
- Arrow: 7.0.0
- Parquet: 7.0.0
- DataFusion: a fork based on 6.0.0

**Proposal**
Maybe you have considered some ideas or solutions about this feature.

**Additional context**

Changelogs
- DataFusion:
  - 7.0.0: https://arrow.apache.org/blog/2022/02/28/datafusion-7.0.0/
  - 8.0.0: https://arrow.apache.org/blog/2022/05/16/datafusion-8.0.0/

Notable Changes:
- DataFusion
  - The `ExecutionPlan` trait is no longer async
  - `RuntimeEnv` is now `TaskContext`, contains more task-relative info like `SessionId` and `TaskId`
  - Bugfix: temp file

### PR DESCRIPTION
**Pull request by waynexia**
Signed-off-by: Ruihang Xia <waynestxia@gmail.com>

# Which issue does this PR close?

Closes #25 

# Rationale for this change
 
 

See #25 

# What changes are included in this PR?

Bump versions.
Target:

- Arrow: 15.0.0
- Parquet: 15.0.0
- DataFusion: 9.0.0 (an unofficial patch branch https://github.com/waynexia/arrow-datafusion/tree/v9.x)

New todo from this PR:
- initialize in `async fn build_scan_table_exec_plan()` should be put off to stream's `poll_next()`
- implement `ContextProvider::get_variable_type()` for `ContextProviderAdapter`
- support `PruningStatistics`'s new method `fn null_counts()` for `RowGroupPruningStatistics`

Significant change:
- rewrite `RecordBytesReader`.
- fix: RecordSchema will ignore ArrowSchema's metadata in https://github.com/CeresDB/ceresdb/pull/26/commits/6b8993c778701f90a6dfc55619a8d45a948d739e

# Are there any user-facing changes?
no
# How does this change test
Unit test

### CODE REVIEW

**Comment by waynexia** [comment]
Change to `refactor` since this touches a lot of logic.

File: components/parquet/src/serialized_reader.rs [hunk 1]
```diff
@@ -232,170 +227,14 @@ impl<'a, R: 'static + ChunkReader> RowGroupReader for SerializedRowGroupReader<'
     }
 }
 
-/// A serialized implementation for Parquet [`PageReader`].
-pub struct SerializedPageReader<T: Read> {
```
Comment (waynexia) [inline components/parquet/src/serialized_reader.rs:236]: This struct is replaced by upstream `SerializedPageReader`

File: query_engine/src/df_execution_extension/prom_align.rs [hunk 2]
```diff
@@ -162,12 +161,17 @@ impl ExecutionPlan for PromAlignExec {
         self.input.output_partitioning()
     }
 
+    // todo: check this ordering
+    fn output_ordering(&self) -> Option<&[PhysicalSortExpr]> {
```
Comment (waynexia) [inline query_engine/src/df_execution_extension/prom_align.rs:165]: @jiacai2050 does `PromAlignExec` guarantee output ordering?

File: analytic_engine/src/sst/parquet/reader.rs [hunk 3]
```diff
@@ -57,20 +58,15 @@ pub async fn read_sst_meta<S: ObjectStore>(
...
-        get_result
-            .bytes()
-            .await
-            .map_err(|e| Box::new(e) as _)
-            .context(ReadPersist {
-                path: path.to_string(),
-            })?
-            .to_vec(),
-    ));
+    let bytes = get_result
```
Comment (ShiKaiWi) [inline analytic_engine/src/sst/parquet/reader.rs:61]: Is the `todo` is solved here?

File: benchmarks/src/parquet_bench.rs [hunk 4]
```diff
@@ -115,7 +114,8 @@ impl ParquetBench {
             }
 
             info!(
-                "\nParquetBench total rows of sst: {}, total batch num: {}, open cost: {:?}, filter cost: {:?}, iter cost: {:?}",
+                "\nParquetBench total rows of sst: {}, total batch num: {},
```
Comment (ShiKaiWi) [inline benchmarks/src/parquet_bench.rs:117]: Remove white space after the colon.

File: query_engine/src/context.rs [hunk 5]
```diff
@@ -69,20 +71,22 @@ impl Builder {
...
+        let logical_optimize_rules = Self::logical_optimize_rules();
+
+        let mut state = default_session_builder(df_exec_config)
+            .with_query_planner(Arc::new(QueryPlannerAdapter))
+            .with_optimizer_rules(logical_optimize_rules);
+        let physical_optimizer =
+            Self::apply_adapters_for_physical_optimize_rules(&state.physical_optimizers);
+        state.physical_optimizers = physical_optimizer;
+
+        let context = SessionContext::with_state(state);
```
Comment (ShiKaiWi) [inline query_engine/src/context.rs:85]: It seems only `context` is needed so how about this style:
```
let df_session_ctx = {
   // building the session ctx
};
```

File: query_engine/src/context.rs [hunk 6]
```diff
@@ -69,20 +71,22 @@ impl Builder {
...
 
         Context {
             request_id: self.request_id,
-            df_exec_ctx: ExecutionContext::with_config(df_exec_config),
+            df_exec_ctx: context,
```
Comment (ShiKaiWi) [inline query_engine/src/context.rs:89]: Rename `df_exec_ctx`  to `df_session_ctx`?

File: sql/src/provider.rs [hunk 7]
```diff
@@ -254,6 +268,14 @@ impl<'a, P: MetaProvider> ContextProvider for ContextProviderAdapter<'a, P> {
             }
         }
     }
+
+    // todo: maybe support this
```
Comment (ShiKaiWi) [inline sql/src/provider.rs:272]: Shall we give more details about this `todo`?

File: table_engine/src/predicate/mod.rs [hunk 8]
```diff
@@ -122,6 +122,11 @@ impl<'a> PruningStatistics for RowGroupPruningStatistics<'a> {
     fn num_containers(&self) -> usize {
         self.row_group_metadata.len()
     }
+
+    // todo: check if this is correct
```
Comment (ShiKaiWi) [inline table_engine/src/predicate/mod.rs:126]: It is right to return None and we can remove this todo.

File: analytic_engine/src/sst/parquet/builder.rs [hunk 9]
```diff
@@ -226,140 +187,97 @@ fn build_write_properties(
...
+            self.fetched_row_num += record_batch.num_rows();
+            arrow_record_batch_vec.push(record_batch.into_record_batch().into_arrow_record_batch());
+
+            if self.fetched_row_num >= self.num_rows_per_row_group {
+                self.fetched_row_num = 0;
+                let row_num = self
+                    .encode_record_batch(&mut arrow_writer, arrow_record_batch_vec)
+                    .map_err(|e| Box::new(e) as _)
+                    .context(EncodeRecordBatch)?;
+                arrow_record_batch_vec = Vec::new();
```
Comment (ShiKaiWi) [inline analytic_engine/src/sst/parquet/builder.rs:214]: Is it better to reserve capacity according to the previous length.

File: analytic_engine/src/sst/parquet/builder.rs [hunk 10]
```diff
@@ -99,16 +94,13 @@ impl std::io::Seek for EncodingBuffer {
...
-    fn try_clone(&self) -> std::io::Result<Self> {
-        Ok(self.clone())
-    }
-}
-
 impl EncodingBuffer {
-    fn read(&self, read_buf: &mut [u8]) -> usize {
-        let mut inner = self.inner.lock().unwrap();
-        inner.read(read_buf)
+    fn into_bytes(self) -> Vec<u8> {
```
Comment (ShiKaiWi) [inline analytic_engine/src/sst/parquet/builder.rs:98]: The `EncodingBuffer` can be simplified (at least no need to wrap vector wth `Arc<Mutex>`) after the `RecordBytesReader` is refactored.

File: query_engine/src/context.rs [hunk 11]
```diff
@@ -54,7 +56,7 @@ pub type ContextRef = Arc<Context>;
 #[must_use]
 pub struct Builder {
     request_id: RequestId,
-    df_exec_config: ExecutionConfig,
+    df_exec_config: SessionConfig,
```
Comment (jiacai2050) [inline query_engine/src/context.rs:59]: df_session_config

File: query_engine/src/context.rs [hunk 12]
```diff
@@ -107,8 +109,7 @@ impl Builder {
             Arc::new(ProjectionPushDown::new()),
             Arc::new(FilterPushDown::new()),
             Arc::new(LimitPushDown::new()),
-            // TODO(xikai): restore this rule after the bug of df is fixed.
-            // Arc::new(SingleDistinctToGroupBy::new()),
+            Arc::new(SingleDistinctToGroupBy::new()),
```
Comment (jiacai2050) [inline query_engine/src/context.rs:112]: Could you link the PR number fixing this bug?

File: query_engine/src/df_planner_extension/mod.rs [hunk 13]
```diff
@@ -25,7 +25,7 @@ impl QueryPlanner for QueryPlannerAdapter {
     async fn create_physical_plan(
         &self,
         logical_plan: &LogicalPlan,
-        ctx_state: &ExecutionContextState,
+        ctx_state: &SessionState,
```
Comment (jiacai2050) [inline query_engine/src/df_planner_extension/mod.rs:28]: session_state

File: query_engine/src/df_planner_extension/prom_align.rs [hunk 14]
```diff
@@ -21,14 +22,15 @@ pub enum Error {
...
 impl ExtensionPlanner for PromAlignPlanner {
-    fn plan_extension(
+    async fn plan_extension(
         &self,
         _planner: &dyn PhysicalPlanner,
         node: &dyn UserDefinedLogicalNode,
         logical_inputs: &[&LogicalPlan],
         physical_inputs: &[Arc<dyn ExecutionPlan>],
-        _ctx_state: &ExecutionContextState,
+        _ctx_state: &SessionState,
```
Comment (jiacai2050) [inline query_engine/src/df_planner_extension/prom_align.rs:33]: session_state

File: query_engine/src/df_planner_extension/table_scan_by_primary_key.rs [hunk 15]
```diff
@@ -88,12 +94,16 @@ impl TableScanByPrimaryKey {
                 // referred to in the query
                 let filters = logical_plan::unnormalize_cols(filters.iter().cloned());
 
-                table_provider.scan_table(
-                    projection,
-                    &filters,
-                    *limit,
-                    ReadOrder::from_is_asc(Some(self.asc)),
-                )
+                // todo: check whether we should push this await down to the result stream.
```
Comment (jiacai2050) [inline query_engine/src/df_planner_extension/table_scan_by_primary_key.rs:97]: I couldn't understand this comment, could you explain this in more details?

File: table_engine/src/predicate/mod.rs [hunk 16]
```diff
@@ -122,6 +122,10 @@ impl<'a> PruningStatistics for RowGroupPruningStatistics<'a> {
     fn num_containers(&self) -> usize {
         self.row_group_metadata.len()
     }
+
+    fn null_counts(&self, _column: &Column) -> Option<ArrayRef> {
+        None
```
Comment (jiacai2050) [inline table_engine/src/predicate/mod.rs:127]: Maybe add a comment to support this later?

File: table_engine/src/provider.rs [hunk 17]
```diff
@@ -60,8 +65,9 @@ impl TableProviderAdapter {
         &self.table
     }
 
-    pub fn scan_table(
+    pub async fn scan_table(
         &self,
+        ctx: &SessionState,
```
Comment (jiacai2050) [inline table_engine/src/provider.rs:70]: state

File: table_engine/src/provider.rs [hunk 18]
```diff
@@ -123,16 +132,45 @@ impl TableProvider for TableProviderAdapter {
 
     async fn scan(
         &self,
+        ctx: &SessionState,
```
Comment (jiacai2050) [inline table_engine/src/provider.rs:135]: state

**Review by jiacai2050 — CHANGES_REQUESTED** [review]

File: query_engine/src/context.rs [hunk 19]
```diff
@@ -54,35 +56,35 @@ pub type ContextRef = Arc<Context>;
...
-            let adapted_physical_optimize_rules = Self::apply_adapters_for_physical_optimize_rules(
-                &self.df_exec_config.physical_optimizers,
-            );
-            let logical_optimize_rules = Self::logical_optimize_rules();
-            self.df_exec_config
-                .with_query_planner(Arc::new(QueryPlannerAdapter))
-                .with_optimizer_rules(logical_optimize_rules)
-                .with_physical_optimizer_rules(adapted_physical_optimize_rules)
-        };
+        let df_exec_config = { self.df_session_config };
```
Comment (ShiKaiWi) [inline query_engine/src/context.rs:74]:
```rust
let df_session_ctx = {
    let mut state = default_session_builder(self.df_session_config)
        .with_query_planner(Arc::new(QueryPlannerAdapter)            
        .with_optimizer_rules(Self::logical_optimize_rules());
    let physical_optimizer = Self::apply_adapters_for_physical_optimize_rules(&state.physical_optimizers);
    state.physical_optimizers = physical_optimizer;
    
    SessionContext::with_state(state)
}
```

**Review by ShiKaiWi — APPROVED** [review]
LGTM

**Review by jiacai2050 — APPROVED** [review]
LGTM

**Pull request merged**